.circleci/config.yml

```diff
@@ -1,241 +1,216 @@
           command: git submodule update --init --recursive
 
-  halt_if_code_unchanged:
-    description: "Halt job if no code changed"
+  store-commit-message:
+    description: "Export Git commit message to environment variable"
+    steps:
+      - run:
+          command: |
+            echo 'export COMMIT_MESSAGE=$(git log -1 --pretty=format:"%s")' >> $BASH_ENV
+            source $BASH_ENV
+          name: "Export COMMIT_MESSAGE environment variable"
+
+  halt-puppeteer-check:
+    description: "Halt if changed file name match ignore patterns"
     steps:
       - run:
+          name: "Halt if changed file name match ignore patterns"
           command: |
-            if ! git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep -qvFf .circleci/e2e-job-ignore-patterns.txt; then
-              echo "Application code have not changed. halting job."
-              circleci step halt
-            else
-              git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep -vFf .circleci/e2e-job-ignore-patterns.txt
-              echo "job continuing."
+            if ! git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep -qvFf .circleci/e2e-job-ignore-patterns.txt ; then
+              echo "Workbench application code are not changed."
+              circleci-agent step halt
+            fi
+      - store-commit-message
+      - run:
+          name: "Halt if commit message contains 'skip e2e'"
+          command: |
+            echo "Git commit message: $COMMIT_MESSAGE"
+            # Double comma is "Parameter Expansion". It converts string to lowercase letters
+            if [[ "${COMMIT_MESSAGE,,}" == *"skip e2e"* ]]; then
+              echo "Halting e2e test."
+              circleci-agent step halt
             fi
-          name: Halt job if no code changed
+      - run:
+          name: "Halt if branch is not a pull request"
+          command: bash .circleci/pr-skip-ci.sh
 
-  ensure_branch_has_changes:
-    description: Halt job and succeed early if no code changes in << parameters.dir_name >> directory on non-master branch
+  halt-test-check:
+    description: Halt job and succeed early if no code changes detected in << parameters.dir_names >> directories on non-master branch
     parameters:
-      dir_name:
-        type: enum
-        enum: ["api", "ui"]
+      dir_names:
+        type: string
+        default: ""
     steps:
       - run:
           command: |
-            if [ ${CIRCLE_BRANCH} != "" ] &&
-              [ ${CIRCLE_BRANCH} != "master" ] &&
-              [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep << parameters.dir_name >>/ | wc -l | xargs) == 0 ]; then
-                echo "No relevant changes in << parameters.dir_name >> directory on non-master branch. halting job."
-                circleci step halt
+            if [ ${CIRCLE_BRANCH} != "" ] && [ ${CIRCLE_BRANCH} != "master" ]; then
```

Comment:
This is a lot of code for an environment we don't really have a way to unit test (to my knowledge; maybe we could spin something up). I kinda want a separate script for this that can be tracked more like code than config.